File: models/sign/sign.go
package sign

import (
	"github.com/merliot/dean"
	"github.com/merliot/hub/models/common"
)

type Display struct {
	Width  int16
	Height int16
}

type Terminal struct {
	Width  int16
	Height int16
}

type Sign struct {
	*common.Common
	Display Display
	Terminal Terminal
	Banner string
	targetStruct
}

var targets = []string{"wioterminal"}

func New(id, model, name string) dean.Thinger {
	println("NEW SIGN")
	s := &Sign{}
	s.Common = common.New(id, model, name, targets).(*common.Common)
	s.targetNew()
	return s
}

func (s *Sign) saveState(msg *dean.Msg) {
	msg.Unmarshal(s).Broadcast()
}

func (s *Sign) getState(msg *dean.Msg) {
	s.Path = "state"
	msg.Marshal(s).Reply()
}

func (s *Sign) save(msg *dean.Msg) {
	msg.Unmarshal(s)
	if s.IsMetal() {
		s.refresh()
		s.store()
	}
	msg.Broadcast()
}

func (s *Sign) Subscribers() dean.Subscribers {
	return dean.Subscribers{
		"state":     s.saveState,
		"get/state": s.getState,
		"save":      s.save,
	}
}

func (s *Sign) Run(i *dean.Injector) {
	s.run(i)
}

File: models/sign/js/sign.js
import { WebSocketController } from './common.js'

export function run(prefix, ws) {
	const sign = new Sign()
	sign.run(prefix, ws)
}

class Sign extends WebSocketController {

	constructor() {
		super()
		this.banner = document.getElementById("banner")
		this.clearBtn = document.getElementById("clear")
		this.saveBtn = document.getElementById("save")

		this.clearBtn.onclick = () => this.clear()
		this.saveBtn.onclick = () => this.save()
	}

	open() {
		super.open()
		this.showBanner()
	}

	handle(msg) {
		switch(msg.Path) {
		case "save":
			this.update(msg)
			break
		}
	}

	clear() {
		this.state.Banner = ""
		this.banner.value = ""
	}

	save() {
		this.state.Banner = this.banner.value
		this.conn.send(JSON.stringify({Path: "save", Banner: this.banner.value}))
	}

	update(msg) {
		this.state.Banner = msg.Banner
		this.banner.value = msg.Banner
	}

	showBanner() {
		let style = window.getComputedStyle(this.banner);
		let oneEm = parseFloat(style.fontSize);
		let heightValue = 1.2 * oneEm * this.state.Terminal.Height;

		this.banner.style.width = this.state.Terminal.Width + 'ch'
		this.banner.style.height = `${heightValue}px`;
		this.banner.value = this.state.Banner
		this.banner.style.display = "block"
	}
}
